src/key_value_store/mod.rs

```diff
@@ -2,73 +2,73 @@ use super::event::Event;
 use core::cell::Cell;
 
 pub trait KeyValueStore<'a> {
-    fn read<T: SafelyDeserializable + Sized>(&self, key: Key) -> T;
-    fn write<T: Sized>(&self, key: Key, value: &T);
+    fn read(&self, key: Key, value: &mut dyn Storable);
+    fn write(&self, key: Key, value: &dyn Storable);
     fn size_of(&self, key: Key) -> Size;
     fn on_change(&self) -> &Event<'a, Key>;
 }
 
-pub trait SafelyDeserializable {
-    fn can_deserialize_from(bytes: &[Cell<u8>]) -> bool;
-    fn size() -> u8;
+pub trait Storable {
+    fn can_deserialize_from(&self, bytes: &[Cell<u8>]) -> bool;
+    fn size(&self) -> u8;
 }
 
-impl SafelyDeserializable for u8 {
-    fn can_deserialize_from(_bytes: &[Cell<u8>]) -> bool {
+impl Storable for u8 {
+    fn can_deserialize_from(&self, _bytes: &[Cell<u8>]) -> bool {
         true
     }
 
-    fn size() -> u8 {
+    fn size(&self) -> u8 {
         1
     }
 }
 
-impl SafelyDeserializable for u16 {
-    fn can_deserialize_from(_bytes: &[Cell<u8>]) -> bool {
+impl Storable for u16 {
+    fn can_deserialize_from(&self, _bytes: &[Cell<u8>]) -> bool {
         true
     }
 
-    fn size() -> u8 {
+    fn size(&self) -> u8 {
         2
     }
 }
 
-impl SafelyDeserializable for u32 {
-    fn can_deserialize_from(_bytes: &[Cell<u8>]) -> bool {
+impl Storable for u32 {
+    fn can_deserialize_from(&self, _bytes: &[Cell<u8>]) -> bool {
         true
     }
 
-    fn size() -> u8 {
+    fn size(&self) -> u8 {
         4
     }
 }
 
-impl SafelyDeserializable for i8 {
-    fn can_deserialize_from(_bytes: &[Cell<u8>]) -> bool {
+impl Storable for i8 {
```

Comment:
I did notice a distinct lack of impl for `bool`, I get why we wouldn't for rust's `char` since it is valid UTF8 and that's a whole bag of worms, but bool should just be:
```rust
impl Storable for bool {
    fn can_deserialize_from(&self, bytes: &[Cell<u8>]) -> bool {
        bytes[0].get() == (false as u8) || bytes[0].get() == (true as u8)
    }

    fn size(&self) -> u8 {
        1
    }
}
```
Is there a reason for not having it, given the simplicity? (similar question for 64 and 128 bit ints)

I might have made a macro to handle the impls so... 
```rust
#[doc(hidden)]
macro_rules! __impl_storable {
    ($typ:ty) => {
        impl Storable for $typ {
            fn can_deserialize_from(&self, _: &[Cell<u8>]) -> bool {
                true
            }

            fn size(&self) -> u8 {
                ::core::mem::size_of::<Self>() as u8
            }
        }
    };
}

__impl_storable!(u8);
__impl_storable!(u16);
__impl_storable!(u32);
__impl_storable!(i8);
__impl_storable!(i16);
__impl_storable!(i32);
```
since `::core::mem::size_of::<Self>() as u8` is a `const` fn, it should be equivalent to just writing the number as far as the compiled code cares

Reply:
Btw, neat note, `Storable`'s `size` fn, could be made a default fn by doing:
```
pub trait Storable {
    fn can_deserialize_from(&self, bytes: &[Cell<u8>]) -> bool;

    fn size(&self) -> u8 {
        core::mem::size_of_val(self) as u8
    }
}
```

Reply:
> I did notice a distinct lack of impl for `bool`

This was omitted for the same reason I omitted `usize`, `isize`, `u64`, `i64`, etc.: I was just being lazy and a couple of `impl`s so I could write some tests while I explored my options.

Reply:
> Btw, neat note, `Storable`'s `size` fn, could be made a default fn by doing:
> 
> ```rust
> pub trait Storable {
>     fn can_deserialize_from(&self, bytes: &[Cell<u8>]) -> bool;
> 
>     fn size(&self) -> u8 {
>         core::mem::size_of_val(self) as u8
>     }
> }
> ```
> 
> which really just means that the `size` fn is un-needed since `size_of_val` will get the correct size for Trait Objects :) (all tests passed when I changed `value.size() == element.size` to `size_of_val(value) as u8 == element.size`)

I didn't realize that traits could have default implementations. Neat!

Reply:
I gave a default implementation for `size`, used the macro, and expanded the `impls` to include other integral and related types,

When I get some time I'll play with `size_of_val`. I'm confused as to how it could possibly work without something like RTTI.

Reply:
I was concerned that `size_of_val` might increase the binary size, but It looks like binary sizes stayed the same (assuming I'm using `cargo size` correctly).